core/src/main/java/jenkins/model/GlobalBuildDiscarderStrategy.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2019 Daniel Beck
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.model;
+
+import hudson.ExtensionPoint;
+import hudson.model.AbstractDescribableImpl;
+import hudson.model.Job;
+import hudson.model.Run;
+
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Extension point for global background build discarders.
+ *
+ * @see BackgroundGlobalBuildDiscarder
+ * @see GlobalBuildDiscarderConfiguration
+ * @see JobGlobalBuildDiscarderStrategy
+ */
+public abstract class GlobalBuildDiscarderStrategy extends AbstractDescribableImpl<GlobalBuildDiscarderStrategy> implements ExtensionPoint {
+    private static final Logger LOGGER = Logger.getLogger(GlobalBuildDiscarderStrategy.class.getName());
+
+    /**
+     * Returns true if and only if this strategy applies to the given job.
+     * @param job
+     * @return true if and only if this strategy applies to the given job.
+     */
+    public abstract boolean isApplicable(Job<?, ?> job);
+
+    /**
+     * Applies this build discarder strategy to the given job, i.e. delete builds based on this strategy's configuration.
+     *
+     * The default implementation calls {@link #apply(Run)} on each build.
+     *
+     * @param job
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    public void apply(Job<? extends Job, ? extends Run> job) throws IOException, InterruptedException {
+        job.getBuilds().forEach(run -> {
+            try {
+                apply(run);
+            } catch (IOException|InterruptedException ex) {
+                // TODO should these actually be caught, or just thrown up to stop applying?
```

Comment:
Whenever you catch an `InterruptedException`, it's usually a good idea to re-set the interrupt flag again via `Thread.currentThread().interrupt();` Otherwise, you can end up with hung threads if they continue running (which in this case, it very well could).

Reply:
So I would not actually want to catch the `InterruptedException` here? The idea was that if there's something wrong with deleting one build's artifacts, e.g. it taking too long to delete a billion files, an initial abort would continue with the next build, rather than just stop this process entirely.

Obviously the downside of that is that it's fairly difficult to really interrupt this thread while it's iterating here.

Reply:
> fairly difficult to really interrupt this thread while it's iterating here.

What processes will intentionally attempt such an interrupt? Is there any jenkins-initiated or user-initiated workflow that can target a specific AsyncPeriodicWork and try to stop it?

Reply:
Setting the interrupt flag doesn't re-throw the interrupted exception btw.

Reply:
I was about to leave a comment about this and noticed that I already did.

Reply:
@daniel-beck Are you planning on making this change? either not catching the InterruptedException or setting the interrupt flag?